cli/src/action/login/oauth/callback.rs

```diff
@@ -0,0 +1,107 @@
+    join_handle: std::thread::JoinHandle<Result<String, OAuth2CallbackError>>,
+}
+
+impl OAuth2Callback {
+    pub fn new() -> Result<Self, OAuth2CallbackError> {
+        let server =
+            Server::http("127.0.0.1:0").map_err(|err| OAuth2CallbackError(err.to_string()))?;
+
+        let port = server.server_addr().port();
+
+        let join_handle = std::thread::Builder::new()
+            .name("Thread-OAuth2Callback".into())
+            .spawn(move || Self::run_callback_server(server))
+            .map_err(|err| OAuth2CallbackError(err.to_string()))?;
+
+        Ok(Self { join_handle, port })
+    }
+
+    pub fn address(&self) -> String {
+        format!("localhost:{}", self.port)
+    }
+
+    pub fn receive_authorization_code(self) -> Result<String, OAuth2CallbackError> {
+        self.join_handle.join().map_err(|_| {
+            OAuth2CallbackError("Unable to join authorization callback thread".into())
+        })?
+    }
+
+    fn run_callback_server(server: Server) -> Result<String, OAuth2CallbackError> {
+        loop {
+            let request = server
+                .recv()
+                .map_err(|err| OAuth2CallbackError(err.to_string()))?;
+
+            if request.method() == &Method::Get && request.url().starts_with("/auth_response") {
+                let code = {
+                    let query = Self::parse_query_str(request.url());
+                    query
+                        .get("code")
+                        .and_then(|vals| vals.iter().next())
+                        .map(|s| s.to_string())
+                };
+
+                if let Some(code) = code {
+                    Self::send_reponse(request)?;
+                    break Ok(code);
+                }
+            } else {
+                debug!("Ignoring request {}: {}", request.method(), request.url());
+            }
+        }
+    }
+
+    fn send_reponse(request: Request) -> Result<(), OAuth2CallbackError> {
+        let response = Response::from_string(include_str!("oauth_success.html")).with_header(
+            Header::from_bytes(&b"Content-Type"[..], &b"text/html"[..])
+                .expect("valid header could not be turned into a header struct"),
+        );
+
+        request
+            .respond(response)
+            .map_err(|err| OAuth2CallbackError(err.to_string()))
+    }
+
+    fn parse_query_str<'a>(uri: &'a str) -> HashMap<&'a str, Vec<&'a str>> {
+        let mut result: HashMap<&'a str, Vec<&'a str>> = HashMap::new();
+
+        let split = uri.split('?');
+        if let Some(query) = split.last() {
+            let params = query.split('&');
+            for param in params {
+                let mut pair = param.split('=');
+                if let Some(key) = pair.next() {
+                    let val = pair.next().unwrap_or("");
+                    result
+                        .entry(key)
+                        .and_modify(|vals| vals.push(val))
+                        .or_insert_with(|| vec![val]);
```

Comment:
It looks like you just take the first value for the `code` key above. Do we need to store multiple values per key? Do we need to get any query values other than `code`?